R/crunch-data-frame.R

```diff
@@ -0,0 +1,292 @@
+    if (is.character(i)) {
+        col_name <- i
+        # check if any don't exist
+    } else if (is.numeric(i)) {
+        col_name <- names(x)[i]
+    } else {
+        halt("column subsetting must be done with either numeric or ",
+             "a character")
+    }
+    return(col_name)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`$.CrunchDataFrame` <- function (x, i) {
+    return(get_CDF_var(i, x, mode = attr(x, "mode")))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`$<-.CrunchDataFrame` <- function (x, i, value) {
+    x <- set_CDF_var(col_name = i, cdf = x, value = value)
+    return(x)
+}
+
+get_CDF_var <- function (col_name, cdf, ...) {
+    if (!is(cdf, "CrunchDataFrame")) { 
+        halt("The cdf argument must be a CrunchDataFrame, got ", class(cdf), " instead.")
+    }
+    if (!col_name %in% names(cdf)) {
+        halt("The variable ", dQuote(col_name), " is not found in the CrunchDataFrame.")
+    }
+    
+    return(get(col_name, cdf)) # need [ord] here?
+}
+
+get_var_from_server <- function (col_name, cdf, ...) {
+    ord <- attr(cdf, "order")
+    ds <- attr(cdf, "crunchDataset")
+    if (!is.null(ord)) {
+        return(
+            as.vector(ds[[col_name]][ord], ...)[match(ord, sort(unique(ord)))]
+        )
+    } else {
+        # no reordering if order is null
+        return(as.vector(ds[[col_name]], ...))
+    }
+}
+
+set_CDF_var <- function (col_name, row_inds, cdf, value) {
+    if (!is(cdf, "CrunchDataFrame")) { 
+        halt("The cdf argument must be a CrunchDataFrame, got ", class(cdf), " instead.")
+    }
+    
+    # remove the variable if the value is NULL
+    if (is.null(value) & missing(row_inds)) {
+        return(invisible(rm_CDF_var(col_name = col_name, cdf = cdf)))
+    }
+    
+    if (missing(row_inds)) {
+        row_inds <- seq_len(nrow(cdf))
+    }
+    
+    # validate row lengths
+    if (length(value) == 1 & length(row_inds) > 1) {
+        # if there is a single value, but more than one row, then we need to 
+        # repeat the value for as many times as there are rows. This is the 
+        # only exception to the rule that there must be the same number of 
+        # values as rows.
+        value <- rep(value, length(row_inds))
+    } else if (is.data.frame(value)) {
+        # we got a data.frame as values, likelythis is an array or MR varaible
+        # unclass does what data.frame(v1=data.frame(a=1, b=2)) would do
+        # value <- unlist(value, recursive = FALSE, use.names = FALSE)
+        # concat names
+        # vnames <- unlist(vnames[ncols > 0L])
+        # set multiple variables
+        value <- I(value)
+    } else if (length(value) != length(row_inds)) {
+        halt("replacement has ", length(value),
+             " rows, the CrunchDataFrame has ", nrow(cdf))
+    }
+    
+    # if we are replacing some rows, fill in with NAs, or grab old vector
+    if (length(row_inds) != nrow(cdf)) {
+        if (!col_name %in% names(cdf)) {
```

Comment:
I personally like to write this as `if (!(col_name %in% names(crdf)))` so that it's clear what is being negated (the whole thing, not `col_name`), despite the fact that it's technically not necessary due to order of operations.

Reply:
👍 I used to almost always do ~this, but at some point have stopped. I'm happy to start again. This might be superstition at this point (and likely doesn't matter at all), but I've read that grouping expressions like this with `{` is actually slightly faster than `(` I also think in this circumstance, it's a bit more readable to have different kinds of braces for finding pairs. `(!{col_name %in% names(crdf)})` is just a bit easier to read for me than `(!(col_name %in% names(crdf)))` 

https://radfordneal.wordpress.com/2010/08/15/two-surpising-things-about-r/